File: source/com/barrybecker4/simulation/snake/SnakeUpdater.java
// Copyright by Barry G. Becker, 2012. Licensed under MIT License: http://www.opensource.org/licenses/MIT
package com.barrybecker4.simulation.snake;

import com.barrybecker4.simulation.snake.geometry.SegmentUpdater;

/**
 *  Makes the snake animate forward one time step.
 *
 *  @author Barry Becker
 */
public class SnakeUpdater {

    private Snake snake;

    /** the time since the start of the simulation */
    private double time_ = 0.0;

    private LocomotionParameters locomotionParams = new LocomotionParameters();
    private SegmentUpdater segmentUpdater;

    /**
     * Constructor
     */
    public SnakeUpdater() {
        segmentUpdater = new SegmentUpdater();
    }

    /**
     * steps the simulation forward in time
     * if the timestep is too big inaccuracy and instability will result.
     * @return the new timestep
     */
    public double stepForward(Snake snake, double timeStep )  {

        this.snake = snake;
        updateParticleForces();

        if ( locomotionParams.getUseFriction() ) {
            updateFrictionalForces();
        }
        updateParticleAccelerations();
        boolean unstable = updateParticleVelocities( timeStep );
        updateParticlePositions( timeStep );

        time_ += timeStep;
        if ( unstable ) {
            return timeStep / 2;
        }
        else {
            return 1.0 * timeStep;
        }
    }

    /**
     * Parameters controlling the movement (i.e. locomotion)
     */
    public LocomotionParameters getLocomotionParams() {
        return locomotionParams;
    }

    /**
     * update forces
     */
    private void updateParticleForces() {

        // apply the sinusoidal muscular contraction function to the
        // left and right sides of the snake
        for ( int i = 2; i <snake.getNumSegments(); i++ )
            snake.getSegment(i).contractMuscles(locomotionParams, time_);

        // update forces based on surrounding contracted springs
        for ( int i = 0; i < snake.getNumSegments(); i++ )
            segmentUpdater.updateForces(snake.getSegment(i));
    }

    /**
     * update accelerations
     */
    private void updateFrictionalForces() {
        for ( int i = 0; i < snake.getNumSegments(); i++ )
            segmentUpdater.updateFrictionalForce(snake.getSegment(i),  locomotionParams);
    }

    /**
     * update accelerations
     */
    private void updateParticleAccelerations() {
        for ( int i = 0; i < snake.getNumSegments(); i++ )
            segmentUpdater.updateAccelerations( snake.getSegment(i));
    }

    /**
     * update velocities
     * @return unstable if velocity changes are getting too big
     */
    private boolean updateParticleVelocities( double timeStep ) {
        boolean unstable = false;
        for ( int i = 0; i < snake.getNumSegments(); i++ )
            if ( segmentUpdater.updateVelocities( snake.getSegment(i), timeStep) )
                unstable = true;
        return unstable;
    }

    /**
     * move particles according to vector field
     */
    private void updateParticlePositions( double timeStep ) {
        for ( int i = 0; i < snake.getNumSegments(); i++ )  {
            segmentUpdater.updatePositions(snake.getSegment(i), timeStep);
        }
    }
}

File: scala-source/com/barrybecker4/simulation/snake/SnakeUpdater.scala
// Copyright by Barry G. Becker, 2016-2017. Licensed under MIT License: http://www.opensource.org/licenses/MIT
package com.barrybecker4.simulation.snake

import com.barrybecker4.simulation.snake.geometry.SegmentUpdater


/**
  * Makes the snake animate forward one time step.
  * @author Barry Becker
  */
class SnakeUpdater private[snake]() {
  private var segmentUpdater = new SegmentUpdater
  private var snake: Snake = _
  /** the time since the start of the simulation */
  private var time = 0.0
  private val locomotionParams = new LocomotionParameters

  /**
    * steps the simulation forward in time
    * if the timestep is too big inaccuracy and instability will result.
    * @return the new timestep
    */
  def stepForward(snake: Snake, timeStep: Double): Double = {
    this.snake = snake
    updateParticleForces()
    if (locomotionParams.useFriction) updateFrictionalForces()
    updateParticleAccelerations()
    val unstable = updateParticleVelocities(timeStep)
    updateParticlePositions(timeStep)
    time += timeStep
    if (unstable) timeStep / 2
    else 1.0 * timeStep
  }

  /**  Parameters controlling the movement (i.e. locomotion) */
  def getLocomotionParams: LocomotionParameters = locomotionParams

  /** update forces */
  private def updateParticleForces(): Unit = { // apply the sinusoidal muscular contraction function to the
    // left and right sides of the snake
    for (i <- 2 until snake.getNumSegments)
      snake.getSegment(i).contractMuscles(locomotionParams, time)

    // update forces based on surrounding contracted springs
    for (i <- 0 until snake.getNumSegments)
      segmentUpdater.updateForces(snake.getSegment(i))
  }

  /** update accelerations */
  private def updateFrictionalForces() {
    for (i <- 0 until snake.getNumSegments)
      segmentUpdater.updateFrictionalForce(snake.getSegment(i), locomotionParams)
  }

  private def updateParticleAccelerations() {
    for (i <- 0 until snake.getNumSegments)
      segmentUpdater.updateAccelerations(snake.getSegment(i))
  }

  /** @return unstable if velocity changes are getting too big */
  private def updateParticleVelocities(timeStep: Double) = {
    var unstable = false
    for (i <- 0 until snake.getNumSegments)
      if (segmentUpdater.updateVelocities(snake.getSegment(i), timeStep)) unstable = true
    unstable
  }

  /** move particles according to vector field */
  private def updateParticlePositions(timeStep: Double) {
    for (i <- 0 until snake.getNumSegments)
      segmentUpdater.updatePositions(snake.getSegment(i), timeStep)
  }
}